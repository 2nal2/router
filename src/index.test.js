import React from "react";
import ReactDOM from "react-dom";
import {
  Router,
  LocationProvider,
  createHistory,
  Link,
  getMatch
} from "./index";
import renderer from "react-test-renderer";
// import Enzyme, { mount } from "enzyme";
// import Adapter from "enzyme-adapter-react-16";
// Enzyme.configure({ adapter: new Adapter() });

const HOME_TEXT = "Home";
const Home = () => <div>{HOME_TEXT}</div>;

const DASH_TEXT = "Dashboard";
const Dash = ({ children }) => (
  <div>
    {DASH_TEXT} {children}
  </div>
);

const REPORTS_TEXT = "Reports";
const Reports = ({ children }) => (
  <div>
    {REPORTS_TEXT} {children}
  </div>
);

const ANNUAL_REPORT_TEXT = "Annual";
const AnnualReport = () => <div>{ANNUAL_REPORT_TEXT}</div>;

const Group = ({ groupId, children }) => (
  <div>
    {groupId} {children}
  </div>
);

////////////////////////////////////////////////////////////////////////////////

const createHistorySource = initialPathname => {
  let index = 0;
  let stack = [{ pathname: initialPathname }];
  let states = [];

  return {
    get location() {
      return stack[index];
    },
    addEventListener(name, fn) {},
    removeEventListener(name, fn) {},
    history: {
      get entries() {
        return stack;
      },
      get index() {
        return index;
      },
      get state() {
        return states[index];
      },
      pushState(state, _, pathname) {
        index++;
        stack.push({ pathname });
        states.push(state);
      },
      replaceState(state, _, pathname) {
        stack[index] = { pathname };
        states[index] = state;
      }
    }
  };
};

const snapshot = ({ pathname, element }) => {
  const testHistory = createHistory(createHistorySource(pathname));
  const wrapper = renderer.create(
    <LocationProvider history={testHistory}>{element}</LocationProvider>
  );
  const tree = wrapper.toJSON();
  expect(tree).toMatchSnapshot();
  return tree;
};

function runWithNavigation(element, pathname = "/") {
  const history = createHistory(createHistorySource(pathname));
  const wrapper = renderer.create(
    <LocationProvider history={history}>{element}</LocationProvider>
  );

  const snapshot = string => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  };

  return { history, snapshot, wrapper };
}

// function runInDOM(element, pathname) {
//   const history = createHistory(createHistorySource(pathname));
//   const wrapper = mount(
//     <LocationProvider history={history}>{element}</LocationProvider>
//   );
//   return { wrapper, history };
// }

////////////////////////////////////////////////////////////////////////////////
// Okay finally some assertions

it("renders a root path component", () => {
  snapshot({
    pathname: "/",
    element: (
      <Router>
        <Home path="/" />
      </Router>
    )
  });
});

it("renders a more specific path", () => {
  snapshot({
    pathname: "/dashboard",
    element: (
      <Router>
        <Home path="/" />
        <Dash path="dashboard" />
      </Router>
    )
  });
});

it("renders a nested path", () => {
  snapshot({
    pathname: "/dash/reports",
    element: (
      <Router>
        <Home path="/" />
        <Dash path="dash">
          <Reports path="reports" />
        </Dash>
      </Router>
    )
  });
});

it("renders a really nested path", () => {
  snapshot({
    pathname: "/dash/reports/annual",
    element: (
      <Router>
        <Home path="/" />
        <Dash path="dash">
          <Reports path="reports">
            <AnnualReport path="annual" />
          </Reports>
        </Dash>
      </Router>
    )
  });
});

it("parses params as component props", () => {
  const USER_ID = "ryan";
  const User = ({ id }) => <div>{id}</div>;

  snapshot({
    pathname: `/users/${USER_ID}`,
    element: (
      <Router>
        <User path="users/:id" />
      </Router>
    )
  });
});

it("parses multiple params as component props", () => {
  const USER_ID = "ryan";
  const GROUP_ID = "party-pooper";
  const User = ({ userId, groupId }) => (
    <div>
      {userId} {groupId}
    </div>
  );

  snapshot({
    pathname: `/${GROUP_ID}/user/${USER_ID}`,
    element: (
      <Router>
        <User path=":groupId/user/:userId" />
      </Router>
    )
  });
});

it("parses multiple params when nested", () => {
  const USER_ID = "ryan";
  const GROUP_ID = "party-pooper";
  const User = ({ userId }) => <div>{userId}</div>;

  snapshot({
    pathname: `/group/${GROUP_ID}/user/${USER_ID}`,
    element: (
      <Router>
        <Group path="group/:groupId">
          <User path="user/:userId" />
        </Group>
      </Router>
    )
  });
});

it("shadows params in nested paths", () => {
  snapshot({
    pathname: `/groups/burger/groups/milkshake`,
    element: (
      <Router>
        <Group path="groups/:groupId">
          <Group path="groups/:groupId" />
        </Group>
      </Router>
    )
  });
});

it("selects a static path over a dynamic segment", () => {
  snapshot({
    pathname: `/home`,
    element: (
      <Router>
        <Group path=":groupId" />
        <Home path="home" />
      </Router>
    )
  });
});

it("selects a static path over a dynamic segment", () => {
  snapshot({
    pathname: `/home`,
    element: (
      <Router>
        <Group path=":groupId" />
        <Home path="home" />
      </Router>
    )
  });
});

it("allows arbitrary Router nesting", () => {
  const Chat = ({ url }) => (
    <div>
      Home
      <ChatApp basepath={url} />
    </div>
  );

  const ChatApp = ({ basepath }) => (
    <Router basepath={basepath}>
      <ChatHome path="/" />
    </Router>
  );

  const ChatHome = ({ url }) => <div>Chat Home</div>;

  snapshot({
    pathname: `/chat`,
    element: (
      <Router>
        <Chat path="/chat" />
      </Router>
    )
  });
});

it("matches on specificity", () => {
  const Groups = ({ children }) => <div>Groups {children}</div>;
  const Group = ({ groupId, children }) => (
    <div>
      Group: {groupId} {children}
    </div>
  );
  const Users = ({ children }) => <div>Users {children}</div>;
  const User = ({ userId, children }) => (
    <div>
      User: {userId} {children}
    </div>
  );

  const Winner = () => <div>Yes I am</div>;

  const element = (
    <Router>
      <Home path="/" />
      <Groups path="groups">
        <Group path=":groupId">
          <Users path="users">
            <User path=":userId" />
          </Users>
        </Group>
      </Groups>
      <Groups path="groups">
        <Group path="mine">
          <Users path="users">
            <User path="me" />
          </Users>
        </Group>
      </Groups>
      <Winner path="groups/gonna/users/win" />
    </Router>
  );

  snapshot({ element, pathname: "/" });

  snapshot({ element, pathname: "/groups" });
  snapshot({ element, pathname: "/groups/123" });
  snapshot({ element, pathname: "/groups/123/users" });
  snapshot({ element, pathname: "/groups/123/users/abc" });
  snapshot({ element, pathname: "/groups/mine/users/me" });

  snapshot({ element, pathname: "/groups/mine" });
  snapshot({ element, pathname: "/groups/mine/users" });
  snapshot({ element, pathname: "/groups/mine/users/me" });

  snapshot({ element, pathname: "/groups/gonna/users/win" });
});

it.only("matches some more", () => {
  const Root = () => <div>Root</div>;
  const Groups = () => <div>Groups</div>;
  const Group = ({ groupId }) => <div>Group Id: {groupId}</div>;
  const MyGroup = () => <div>MyGroup</div>;
  const MyGroupsUsers = () => <div>MyGroupUsers</div>;
  const Users = () => <div>Users</div>;
  const UsersSplat = ({ splat }) => <div>Users Splat: {splat}</div>;
  const UsersOptional = ({ userId }) => (
    <div>Users Optional: {userId}</div>
  );
  const User = ({ userId, groupId }) => (
    <div>
      User id: {userId}, Group Id: {groupId}
    </div>
  );
  const Me = () => <div>Me!</div>;
  const MyGroupsAndMe = () => <div>Mine and Me!</div>;
  const Fiver = ({ one, two, three, four, five }) => (
    <div>
      Fiver {one} {two} {three} {four} {five}
    </div>
  );

  const element = (
    <Router>
      <Root path="/" />
      <Groups path="/groups" />
      <Group path="/groups/:groupId" />
      <MyGroup path="/groups/mine" />
      <Users path="/groups/:groupId/users" />
      <MyGroupsUsers path="/groups/mine/users" />
      <UsersSplat path="/groups/:groupId/users/:splat*" />
      <UsersOptional path="/groups/:groupId/users/:userId?" />
      <User path="/groups/:groupId/users/:userId" />
      <Me path="/groups/:groupId/users/me" />
      <MyGroupsAndMe path="/groups/:groupId/users/me" />
      <Fiver path="/:one/:two/:three/:four/:five*" />
    </Router>
  );

  snapshot({ element, pathname: "/" }); // Root
  snapshot({ element, pathname: "/groups" }); // Groups
  snapshot({ element, pathname: "/groups/123" }); // Group
  snapshot({ element, pathname: "/groups/mine" }); // MyGroup
  snapshot({ element, pathname: "/groups/123/users" }); // Users
  snapshot({ element, pathname: "/groups/mine/users" }); // MyGroupsUsers
  snapshot({ element, pathname: "/groups/123/users/456" }); // User
  snapshot({ element, pathname: "/groups/123/users/me" }); // Me
  snapshot({ element, pathname: "/groups/123/users/a/bunch/of/junk" }); // UsersSplat
  snapshot({ element, pathname: "/groups/mine/users/me" }); // MyGroupsAndMe
  snapshot({ element, pathname: "/one/two/three/four/five" }); // Fiver
});

it("transitions pages", async () => {
  const { snapshot, history: { navigate } } = runWithNavigation(
    <Router>
      <Home path="/" />
      <Reports path="reports" />
    </Router>
  );
  snapshot();
  await navigate("/reports");
  snapshot();
});

it("keeps the stack right on interrupted transitions", async () => {
  const {
    snapshot,
    history,
    history: { navigate }
  } = runWithNavigation(
    <Router>
      <Home path="/" />
      <Reports path="reports" />
      <AnnualReport path="annual-report" />
    </Router>
  );
  navigate("/reports");
  await navigate("/annual-report");
  snapshot();
  expect(history.index === 1);
});

it.skip("supports relative links", async () => {
  // something is up with jsdom (maybe), relative links work in the browser,
  // but the href doesn't resolve correctly in here, will look into later
  // const Reports = () => (
  //   <div>
  //     <Link to="annual-report">Annual Report</Link>
  //   </div>
  // );
  // const { wrapper, history } = runInDOM(
  //   <Router>
  //     <Home path="/" />
  //     <Reports path="reports">
  //       <AnnualReport path="annual-report" />
  //     </Reports>
  //   </Router>,
  //   "/reports"
  // );
  // expect(wrapper.html()).toEqual(
  //   '<div><a href="annual-report">Annual Report</a></div>'
  // );
  // await new Promise(res => {
  //   history.listen(res);
  //   wrapper.find("a").simulate("click", { button: 0 });
  // });
  // expect(wrapper.html()).toEqual(
  //   '<div><a href="annual-report">Annual Report</a></div>'
  // );
});

it("matches", () => {
  getMatch({ pathname: "/foo/bar/" });
});

/*

/groups/mine
/groups/mine/users
/groups/mine/users/me
/groups/:groupId
/groups/:groupId/users
/groups/:groupId/users/me
/groups/:groupId/users/:splat*
/groups/:groupId/users/:userId

*/
