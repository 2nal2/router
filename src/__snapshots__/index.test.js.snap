// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchPath matches a path 1`] = `
<pre>
  {
  "location": {
    "pathname": "/groups/123"
  },
  "match": {
    "groupId": "123",
    "uri": "/groups/123",
    "path": "/groups/:groupId"
  }
}
</pre>
`;

exports[`disrespect has complete disrespect for leading and trailing slashes 1`] = `
<div>
  Dash 
  <div>
    Reports 
    <div>
      Annual Report
    </div>
  </div>
</div>
`;

exports[`links renders links with relative hrefs 1`] = `
<div>
  <h1>
    Parent
  </h1>
  <a
    href="/dash/reports"
    location={
      Object {
        "pathname": "/dash/reports",
      }
    }
    onClick={[Function]}
  >
    /dash/reports
  </a>
  <div>
    <h2>
      Child
    </h2>
    <a
      href="/dash"
      location={
        Object {
          "pathname": "/dash/reports",
        }
      }
      onClick={[Function]}
    >
      /dash
    </a>
  </div>
</div>
`;

exports[`nested rendering matches multiple nested / down to a child with a path 1`] = `
<div>
  Dash 
  <div>
    Dash 
    <div>
      Dash 
    </div>
  </div>
</div>
`;

exports[`nested rendering renders a child 'index' nested path 1`] = `
<div>
  Dash 
  <div>
    Reports 
  </div>
</div>
`;

exports[`nested rendering renders a nested path 1`] = `
<div>
  Dash 
  <div>
    Reports 
  </div>
</div>
`;

exports[`nested rendering renders a really nested path 1`] = `
<div>
  Dash 
  <div>
    Reports 
    <div>
      Annual Report
    </div>
  </div>
</div>
`;

exports[`nested rendering yo dawg 1`] = `
<div>
  Dash 
  <div>
    Dash 
    <div>
      Dash 
    </div>
  </div>
</div>
`;

exports[`nested routers allows arbitrary Router nesting through context 1`] = `
<div>
  Home
  <div>
    Chat Home
  </div>
</div>
`;

exports[`passed props parses dynamic segments and passes to components 1`] = `
<div>
  Group: 
  123
</div>
`;

exports[`passed props parses multiple params when nested 1`] = `
<div>
  123
  <div>
    123
     - 
    456
  </div>
</div>
`;

exports[`passed props passes the matched URI to the component 1`] = `
<pre>
  {
  "path": "/groups/:groupId/users/:userId",
  "groupId": "123",
  "userId": "456",
  "uri": "/groups/123/users/456",
  "location": {
    "pathname": "/groups/123/users/456"
  }
}
</pre>
`;

exports[`passed props shadows params in nested paths 1`] = `
<div>
  Group: 
  burger
  <div>
    Group: 
    milkshake
  </div>
</div>
`;

exports[`route ranking / 1`] = `
<div>
  Root
</div>
`;

exports[`route ranking /groups 1`] = `
<div>
  Groups
</div>
`;

exports[`route ranking /groups/123 1`] = `
<div>
  Group Id: 
  123
</div>
`;

exports[`route ranking /groups/123/users 1`] = `
<div>
  Users
</div>
`;

exports[`route ranking /groups/123/users/456 1`] = `
<div>
  User id: 
  456
  , Group Id: 
  123
</div>
`;

exports[`route ranking /groups/123/users/a/bunch/of/junk 1`] = `
<div>
  Users Splat: 
</div>
`;

exports[`route ranking /groups/123/users/me 1`] = `
<div>
  Me!
</div>
`;

exports[`route ranking /groups/mine 1`] = `
<div>
  MyGroup
</div>
`;

exports[`route ranking /groups/mine/users 1`] = `
<div>
  MyGroupUsers
</div>
`;

exports[`route ranking /groups/mine/users/me 1`] = `
<div>
  Mine and Me!
</div>
`;

exports[`route ranking /one/two/three/four/five 1`] = `
<div>
  Fiver 
  one
   
  two
   
  three
   
  four
   
  five
</div>
`;

exports[`smoke tests renders at a path 1`] = `
<div>
  Dash 
</div>
`;

exports[`smoke tests renders the root component at "/" 1`] = `
<div>
  Home
</div>
`;

exports[`transitions keeps the stack right on interrupted transitions 1`] = `
<div>
  Annual Report
</div>
`;

exports[`transitions transitions pages 1`] = `
<div>
  Home
</div>
`;

exports[`transitions transitions pages 2`] = `
<div>
  Reports 
</div>
`;
